File: packages/sdk/src/b64.ts
// `b64:` + standard padded base64, matching apps/backend/src/auth/codec.mjs
// exactly (docs/contracts/crypto-envelope-v1.md: "Binary API value |
// `b64:` + standard RFC 4648 base64 with padding").
const PREFIX = "b64:";

export function encodeB64(bytes: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < bytes.length; i += 1) binary += String.fromCharCode(bytes[i]);
  return PREFIX + btoa(binary);
}

export function decodeB64(value: string): Uint8Array {
  if (!value.startsWith(PREFIX)) {
    throw new Error(`expected a '${PREFIX}'-prefixed base64 string`);
  }
  const binary = atob(value.slice(PREFIX.length));
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i += 1) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

File: apps/android/app/src/main/kotlin/com/pass/android/codec/B64.kt
package com.pass.android.codec

import java.util.Base64

/** `b64:` + standard padded base64, matching `packages/sdk/src/b64.ts` and
 * `apps/backend/src/auth/codec.mjs` exactly (docs/contracts/crypto-envelope-v1.md:
 * "Binary API value | `b64:` + standard RFC 4648 base64 with padding"). */
private const val PREFIX = "b64:"

fun encodeB64(bytes: ByteArray): String = PREFIX + Base64.getEncoder().encodeToString(bytes)

class InvalidB64Exception : Exception("expected a 'b64:'-prefixed base64 string")

fun decodeB64(value: String): ByteArray {
    if (!value.startsWith(PREFIX)) throw InvalidB64Exception()
    return try {
        Base64.getDecoder().decode(value.substring(PREFIX.length))
    } catch (e: IllegalArgumentException) {
        throw InvalidB64Exception()
    }
}
